tools/src/main/scala/scala/scalanative/build/NativeSourcePlugin.scala

```diff
@@ -0,0 +1,256 @@
+
+import java.io.File
+import java.nio.file.{Files, Path, Paths}
+import scalanative.build.IO.RichPath
+import scalanative.build.Mode._
+
+import NativeSourcePlugin._
+
+trait NativeSourcePlugin {
+
+  def extensions: Seq[String]
+
+  def compile(config: Config, path: Path): Option[CompilationContext]
+
+}
+
+object NativeSourcePlugin {
+
+  case class CompilationContext(
+      command: Seq[String],
+      result: CompilationResult
+  )
+
+}
+
+sealed abstract class ClangSourcePlugin extends NativeSourcePlugin {
+
+  protected def stdflag(config: Config): Seq[String]
+
+  protected def compiler(config: Config): String = config.clang.abs
+
+  protected def forceRebuild: Boolean = false
+
+  private def opt(config: Config): String =
+    config.mode match {
+      case Mode.Debug       => "-O0"
+      case Mode.ReleaseFast => "-O2"
+      case Mode.ReleaseFull => "-O3"
+    }
+
+  def compile(config: Config, inpath: Path): Option[CompilationContext] = {
+    val outpath = Paths.get(inpath.abs + LLVM.oExt)
+    val alreadyExists = Files.exists(outpath)
+    if (forceRebuild || !alreadyExists) Some {
+      val platformFlags = {
+        if (config.targetsWindows) Seq("-g")
+        else Nil
+      }
+      val flags = opt(config) +: "-fvisibility=hidden" +:
+        stdflag(config) ++: platformFlags ++: config.compileOptions
+
+      CompilationContext(
+        command = Seq(compiler(config)) ++
+          LLVM.flto(config) ++
+          flags ++
+          LLVM.target(config) ++
+          Seq("-c", inpath.abs, "-o", outpath.abs),
+        result = ObjectFile(outpath)
+      )
+    }
+    else if (alreadyExists) Some(CompilationContext(Nil, ObjectFile(outpath)))
+    else None
+  }
+
+}
+
+case object LlSourcePlugin extends ClangSourcePlugin {
+
+  /** LLVM intermediate file extension: ".ll" */
+  val llExt = ".ll"
+
+  val extensions = Seq(llExt)
+
+  // LL is generated so always rebuild
+  override def forceRebuild = true
+
+  protected def stdflag(config: Config): Seq[String] = Seq()
+
+}
+
+case object CSourcePlugin extends ClangSourcePlugin {
+
+  /** C file extension: ".c" */
+  val cExt = ".c"
+
+  /** Assembler file extension: ".S" */
+  val assemblerExt = ".S"
+
+  /** List of source patterns used: ".c, .cpp, .S" */
+  val extensions = Seq(cExt, assemblerExt)
+
+  protected def stdflag(config: Config): Seq[String] = Seq("-std=gnu11")
+
+}
+
+case object CppSourcePlugin extends ClangSourcePlugin {
+
+  /** C++ file extension: ".cpp" */
+  val cppExt = ".cpp"
+
+  val extensions = Seq(cppExt)
+
+  protected def stdflag(config: Config): Seq[String] =
+    // C++14 or newer standard is needed to compile code using Windows API
+    // shipped with Windows 10 / Server 2016+ (we do not plan supporting older versions)
+    if (config.targetsWindows) Seq("-std=c++14")
+    else Seq("-std=c++11")
+
+  protected override def compiler(config: Config): String = config.clangPP.abs
+
+}
+
+case object RustSourcePlugin extends NativeSourcePlugin {
```

Comment:
I kept this here for now, but I guess the idea is this can move into another repo/library? :)

Reply:
Yes, I agree :) It can be complicated to modify the CI in the primary repo to accomodate rust's special requirements. I think all the rust integration can live very naturally in another repo: a compiler plugin, some std library wrappers like you suggested, etc.